frontends/concrete-python/concrete/fhe/mlir/processors/assign_bit_widths.py

```diff
@@ -13,266 +13,346 @@
-
-    Nodes are just integers id.
-    """
+            bit_width (z3.Int):
+                symbolic bit-width which will be assigned to node once constraints are solved
+        """
 
-    parent: list[int]
-
-    def __init__(self, size: int):
-        """Create a union find suitable for `size` nodes."""
-        self.parent = list(range(size))
-
-    @property
-    def size(self):
-        """Size in number of nodes."""
-        return len(self.parent)
-
-    def find_canonical(self, a: int) -> int:
-        """Find the current canonical node for a given input node."""
-        parent = self.parent[a]
-        if a == parent:
-            return a
-        canonical = self.find_canonical(parent)
-        self.parent[a] = canonical
-        return canonical
-
-    def union(self, a: int, b: int):
-        """Union both nodes."""
-        self.united_common_ancestor(a, b)
-
-    def united_common_ancestor(self, a: int, b: int) -> int:
-        """Deduce the common ancestor of both nodes after unification."""
-        parent_a = self.parent[a]
-        parent_b = self.parent[b]
-
-        if parent_a == parent_b:
-            return parent_a
-
-        if a == parent_a and parent_b < parent_a:
-            common_ancestor = parent_b
-        elif b == parent_b and parent_a < parent_b:
-            common_ancestor = parent_a
-        else:
-            common_ancestor = self.united_common_ancestor(parent_a, parent_b)
-
-        self.parent[a] = common_ancestor
-        self.parent[b] = common_ancestor
-        return common_ancestor
+        assert node.operation in {Operation.Generic, Operation.Constant, Operation.Input}
+        operation = (
+            node.properties["name"]
+            if node.operation == Operation.Generic
+            else ("constant" if node.operation == Operation.Constant else "input")
+        )
+
+        if hasattr(self, operation):
+            self.node = node
+            self.bit_width = bit_width
+
+            try:
+                preds = self.graph.ordered_preds_of(node)
+                constraints = getattr(self, operation)
+
+                if isinstance(constraints, set):
+                    for constraint in constraints:
+                        constraint(self, preds)
+
+                elif isinstance(constraints, dict):
+                    for condition, conditional_constraints in constraints.items():
+                        if condition(self, preds):
+                            for constraint in conditional_constraints:
+                                constraint(self, preds)
+
+                else:  # pragma: no cover
+                    message = (
+                        f"Expected a set or a dict "
+                        f"for additional constraints of '{operation}' operation"
+                        f"but got {type(constraints).__name__} instead"
+                    )
+                    raise ValueError(message)
+
+            finally:
```

Comment:
I think having them as parameter is better.
Keep in mind that it could be just 1 parameter.